x-pack/protocol/src/main/java/org/elasticsearch/protocol/xpack/XPackInfoResponse.java

```diff
@@ -0,0 +1,499 @@
+    public void readFrom(StreamInput in) throws IOException {
+        this.buildInfo = in.readOptionalWriteable(BuildInfo::new);
+        this.licenseInfo = in.readOptionalWriteable(LicenseInfo::new);
+        this.featureSetsInfo = in.readOptionalWriteable(FeatureSetsInfo::new);
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        if (other == null || other.getClass() != getClass()) return false;
+        if (this == other) return true;
+        XPackInfoResponse rhs = (XPackInfoResponse) other;
+        return Objects.equals(buildInfo, rhs.buildInfo)
+                && Objects.equals(licenseInfo, rhs.licenseInfo)
+                && Objects.equals(featureSetsInfo, rhs.featureSetsInfo);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(buildInfo, licenseInfo, featureSetsInfo);
+    }
+
+    @Override
+    public String toString() {
+        return Strings.toString(this, true, false);
+    }
+
+    private static final ConstructingObjectParser<XPackInfoResponse, Void> PARSER = new ConstructingObjectParser<>(
+            "xpack_info_response", true, (a, v) -> {
+                BuildInfo buildInfo = (BuildInfo) a[0];
+                LicenseInfo licenseInfo = (LicenseInfo) a[1];
+                @SuppressWarnings("unchecked") // This is how constructing object parser works
+                List<FeatureSetsInfo.FeatureSet> featureSets = (List<FeatureSetsInfo.FeatureSet>) a[2];
+                FeatureSetsInfo featureSetsInfo = featureSets == null ? null : new FeatureSetsInfo(new HashSet<>(featureSets));
+                return new XPackInfoResponse(buildInfo, licenseInfo, featureSetsInfo);
+            });
+    static {
+        PARSER.declareObject(optionalConstructorArg(), BuildInfo.PARSER, new ParseField("build"));
+        /*
+         * licenseInfo is sort of "double optional" because it is
+         * optional but it can also be send as `null`.
+         */
+        PARSER.declareField(optionalConstructorArg(), (p, v) -> {
+                    if (p.currentToken() == XContentParser.Token.VALUE_NULL) {
+                        return null;
+                    }
+                    return LicenseInfo.PARSER.parse(p, v);
+                },
+                new ParseField("license"), ValueType.OBJECT_OR_NULL);
+        PARSER.declareNamedObjects(optionalConstructorArg(),
+                (p, c, name) -> FeatureSetsInfo.FeatureSet.PARSER.parse(p, name),
+                new ParseField("features"));
+    }
+    public static XPackInfoResponse fromXContent(XContentParser parser) throws IOException {
+        return PARSER.parse(parser, null);
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject();
+
+        if (buildInfo != null) {
+            builder.field("build", buildInfo, params);
+        }
+
+        EnumSet<XPackInfoRequest.Category> categories = XPackInfoRequest.Category
```

Comment:
Before this change we'd read this information from the `XPackInfoRequest` object. With this change we parse it twice. I don't think that is a big loss though because this API isn't performance critical and `toXContent` can't see the request anyway.